{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the gene file +/- 1000 base pairs and break it up into 3 pieces\n",
    "def read_gene_plus_string(Gene_plus):\n",
    "\n",
    "    # # Import Gene -+1000 base pairs - clean in the same way\n",
    "    # with open(genefile, 'r') as file:\n",
    "    #     Gene_plus = file.read().replace('\\n', '')\n",
    "\n",
    "    # Clean away numbers and spaces\n",
    "    Gene_plus = Gene_plus.replace(' ', '') # spaces\n",
    "    remove_digits = str.maketrans('', '', '0123456789')\n",
    "    Gene_plus = Gene_plus.translate(remove_digits)   # numbers\n",
    "    Left_of_gene = Gene_plus[0:1000] # 1000 bp's to the left of the Gene\n",
    "    Right_of_gene = Gene_plus[(len(Gene_plus)-1000):] # 1000 bp's to the right of the Gene (includes 3' UTR segment)\n",
    "    Gene = Gene_plus[1000:(len(Gene_plus)-1000)]\n",
    "    \n",
    "    return (Left_of_gene, Gene, Right_of_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in and process the enzyme list\n",
    "def read_enzyme_list(enzymefile):\n",
    "\n",
    "    with open(enzymefile) as file:\n",
    "        lines=file.readlines()\n",
    "        enzymes=[line.rstrip() for line in lines]\n",
    "        \n",
    "    rsitelist = []\n",
    "    enamelist = []\n",
    "    if (len(enzymes) % 2) == 1:\n",
    "        print('Enzyme list has an odd number of lines. Enzyme list should be a list of enzyme names and restriction sites in each line.')\n",
    "    else:\n",
    "        for linei in range(1, len(enzymes), 2):\n",
    "            rsite = enzymes[linei]\n",
    "            ename = enzymes[linei-1]\n",
    "            # check length of restriction site\n",
    "            if len(rsite) == 6 or len(rsite) == 8:\n",
    "                # check whether any non-ACGT characters\n",
    "                if len(re.sub('[ACGT]', '', rsite)) == 0:\n",
    "                    #check whether palindromic so can focus on only one strand\n",
    "                    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "                    reverse_complement = \"\".join(complement.get(base, base) for base in reversed(rsite))\n",
    "                    if rsite==reverse_complement:\n",
    "                        # add name only if enzyme already in the list\n",
    "                        rsiteexists=0 \n",
    "                        for linej in range(len(rsitelist)):\n",
    "                            if rsitelist[linej] == rsite:\n",
    "                                rsiteexists=1\n",
    "                                enamelist[linej] = enamelist[linej]+', '+ename\n",
    "                                break\n",
    "                        #add new entry if not yet there\n",
    "                        if rsiteexists==0:\n",
    "                            rsitelist.append(rsite)\n",
    "                            enamelist.append(ename)\n",
    "    return (np.array(rsitelist), np.array(enamelist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genefile   = './Chris_code/clb2_pm_1000.txt'\n",
    "# enzymefile = './Chris_code/raw_enzyme_list.txt'\n",
    "# minhomology = 100\n",
    "# alpha = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Left_of_gene, Gene, Right_of_gene = read_gene_string(genefile)\n",
    "# rsitelist, enamelist = read_enzyme_list(enzymefile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_rsitelist_sorted, gene_enamelist_sorted, gene_rsite_position_list_sorted = right_search(Gene, rsitelist, enamelist, minhomology=100)\n",
    "# gene_rsitelist_sorted, gene_enamelist_sorted, gene_rsite_position_list_sorted"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4. and 5."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # 4. and 5.\n",
    "\n",
    "# ind = 1\n",
    "\n",
    "# rsite0 = gene_rsitelist_sorted[ind]\n",
    "# rsite0_pos = gene_rsite_position_list_sorted[ind]\n",
    "# ename0 = gene_enamelist_sorted[ind]\n",
    "# X = Gene[rsite0_pos-minhomology:-3]\n",
    "# alphaX = Right_of_gene[:int(alpha*len(X))]\n",
    "\n",
    "\n",
    "# X.find(rsite0)==minhomology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "# linker = load_linker('')\n",
    "# FPG = load_FPG('')\n",
    "\n",
    "# full_sequence = X + linker + FPG + alphaX\n",
    "# if full_sequence.count(rsite0) != 1:\n",
    "#     print('rsite is not unique in the gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 7."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MCS = load_MCS('')\n",
    "# # Search for the most outer CS from MCS that are not in 4. \n",
    "\n",
    "# rsite1 = ''\n",
    "# rsite2 = ''\n",
    "\n",
    "# msc_rsitelist_sorted_right, _, _ = left_search(MCS, rsitelist, enamelist)\n",
    "# for i, rsite in enumerate(msc_rsitelist_sorted_right):\n",
    "#     if rsite not in full_sequence:\n",
    "#         rsite1 = rsite\n",
    "#         break\n",
    "\n",
    "# msc_rsitelist_sorted_left, _, _ = right_search(MCS, rsitelist, enamelist)\n",
    "# for i, rsite in enumerate(msc_rsitelist_sorted_left):\n",
    "#     if rsite not in full_sequence:\n",
    "#         rsite2 = rsite\n",
    "#         break\n",
    "\n",
    "# if not rsite1 or not rsite2:\n",
    "#     print('good rsites not found in MCS')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# def generate_full_tag_sequence(rsite_chunk, prime_5_chunk, prime_3_chunk, rsite_side, rsite_position, rsite, minhomology, alpha, linker, FPG):\n",
    "\n",
    "#     #TODO 5p i 3p with respect to the main gene sequence\n",
    "\n",
    "#     if rsite_side not in [3,5]:\n",
    "#         raise ValueError('Wrong Gene_side value')\n",
    "\n",
    "#     if rsite_side == 5:\n",
    "#         X = rsite_chunk[rsite_position-minhomology:-3] #We retrieve the 5p end of the gene without the stop codon\n",
    "#         alphaX = prime_3_chunk[:int(alpha*len(X))]\n",
    "#         full_sequence = X + linker + FPG + alphaX + 'TAA' #we add the stop codon at the end\n",
    "#     elif rsite_side == 3:\n",
    "#         # TODO check\n",
    "#         X = rsite_chunk[:rsite_position+len(rsite)+minhomology]\n",
    "#         alphaX = prime_5_chunk[-int(alpha*len(X)):]\n",
    "#         full_sequence = alphaX + FPG + linker + X \n",
    "\n",
    "#     return full_sequence#, X, alphaX\n",
    "\n",
    "# def generate_full_3UTR_sequence(Gene, Right_of_gene, rsite_position, rsite, minhomology, alpha, linker, FPG):\n",
    "\n",
    "#     return generate_full_tag_sequence(\n",
    "#         rsite_chunk=Right_of_gene,\n",
    "#         prime_5_chunk=Gene,\n",
    "#         prime_3_chunk='', \n",
    "#         rsite_side=3, \n",
    "#         rsite_position=rsite_position, \n",
    "#         rsite=rsite, \n",
    "#         minhomology=minhomology, \n",
    "#         alpha=alpha, \n",
    "#         linker=linker, \n",
    "#         FPG=FPG\n",
    "#     )\n",
    "\n",
    "# def generate_full_5UTR_sequence(Gene, Left_of_gene, rsite_position, rsite, minhomology, alpha, linker, FPG):\n",
    "\n",
    "#     return generate_full_tag_sequence(\n",
    "#         rsite_chunk=Left_of_gene,\n",
    "#         prime_3_chunk=Gene, \n",
    "#         prime_5_chunk='',\n",
    "#         rsite_side=5, \n",
    "#         rsite_position=rsite_position, \n",
    "#         rsite=rsite, \n",
    "#         minhomology=minhomology, \n",
    "#         alpha=alpha, \n",
    "#         linker=linker, \n",
    "#         FPG=FPG\n",
    "#     )\n",
    "\n",
    "# def generate_full_delete_sequence(Left_of_gene, Right_of_gene, rsite_side, rsite_position, rsite, minhomology, alpha):\n",
    "\n",
    "#     #TODO 5p i 3p with respect to the main gene sequence\n",
    "\n",
    "#     if rsite_side not in [3,5]:\n",
    "#         raise ValueError('Wrong rsite_side value')\n",
    "#     if rsite_side == 5:\n",
    "#         X = Left_of_gene[rsite_position-minhomology:]\n",
    "#         alphaX = Right_of_gene[-int(alpha*len(X)):]\n",
    "#         full_sequence = X + alphaX\n",
    "#     elif rsite_side == 3:\n",
    "#         X = Right_of_gene[:rsite_position+len(rsite)+minhomology]\n",
    "#         alphaX = Left_of_gene[-int(alpha*len(X)):]\n",
    "#         full_sequence = alphaX + X\n",
    "\n",
    "#     return full_sequence#, X, alphaX\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 8."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "def assemble_plasmid(backbone_no_MCS_5, backbone_no_MCS_3, sequence): ## TODO CHECK if backbone_no_MCS_5 + sequence + backbone_no_MCS_3 + backbone_no_MCS_5 bad\n",
    "    return backbone_no_MCS_5 + sequence + backbone_no_MCS_3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FULL CODE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_from_fsa(fsa_file_path):\n",
    "    if not fsa_file_path:\n",
    "        return '',''\n",
    "    try:\n",
    "        with open(fsa_file_path) as f:\n",
    "            name = f.readline()[1:-1]\n",
    "            code = f.read().replace('\\n', '')\n",
    "            return name, code.upper()\n",
    "    except Exception as e:\n",
    "        print('Unable to read:', fsa_file_path,e)\n",
    "\n",
    "# read_from_fsa('./Test_examples/pETUL_backbone.fsa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3.\n",
    "\n",
    "def rsite_search_func(Gene, rsitelist, enamelist, starting_end, min_homology=0, \n",
    "                max_homology=1000):\n",
    "\n",
    "    if starting_end == 5:\n",
    "        Gene_cut = Gene[min_homology:]\n",
    "        rsite_position_list = np.array([Gene_cut.find(rsite) for rsite in rsitelist])\n",
    "        sorted_inds = np.argsort(rsite_position_list)\n",
    "        \n",
    "    elif starting_end == 3:\n",
    "        Gene_cut = Gene[len(Gene)-max_homology:len(Gene)-min_homology]\n",
    "        rsite_position_list = np.array([Gene_cut.rfind(rsite) for rsite in rsitelist]) \n",
    "        sorted_inds = np.argsort(-rsite_position_list)\n",
    "\n",
    "\n",
    "    # remove not matched rsites\n",
    "    rsite_position_list = rsite_position_list[sorted_inds]\n",
    "    matched_inds = (rsite_position_list != -1)\n",
    "    rsite_position_list = rsite_position_list[matched_inds]\n",
    "    if starting_end == 5:\n",
    "        rsite_position_list += min_homology\n",
    "\n",
    "    # apply sort and remove inds to rsitelist and enamelist\n",
    "    rsitelist = rsitelist[sorted_inds][matched_inds]\n",
    "    enamelist = enamelist[sorted_inds][matched_inds]\n",
    "\n",
    "    return rsitelist, enamelist, rsite_position_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 468,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_start_end_sequences(left_chunk, right_chunk, rsite_side, rsite_position, rsite, minhomology, alpha):\n",
    "\n",
    "    if rsite_side not in ['left','right']:\n",
    "        raise ValueError('Wrong Gene_side value')\n",
    "\n",
    "    if rsite_side == 'left':\n",
    "        X = left_chunk[rsite_position-minhomology:-3]\n",
    "        alphaX = right_chunk[:int(alpha*len(X))]\n",
    "        return X, alphaX\n",
    "    else:\n",
    "        X = right_chunk[:rsite_position+len(rsite)+minhomology]\n",
    "        alphaX = left_chunk[-int(alpha*len(X))-3:-3]\n",
    "        return alphaX, X\n",
    "\n",
    "def find_compatible_MCS_rsites(MCS, rsitelist, enamelist, full_sequences, backbone_no_MCS_5, backbone_no_MCS_3):\n",
    "\n",
    "    rsite1 = ''\n",
    "    ename1 = ''\n",
    "    rsite2 = ''\n",
    "    ename2 = ''\n",
    "\n",
    "    backbone_no_MCS = backbone_no_MCS_3 + backbone_no_MCS_5\n",
    "\n",
    "    rsitelist_sorted_right, enamelist_sorted_right, _ = rsite_search_func(MCS, rsitelist, enamelist, starting_end=5)\n",
    "    for i, rsite in enumerate(rsitelist_sorted_right):\n",
    "        if not any(rsite in full_sequence for full_sequence in full_sequences) and not rsite in backbone_no_MCS:\n",
    "            rsite1 = rsite\n",
    "            ename1 = enamelist_sorted_right[i]\n",
    "            break\n",
    "\n",
    "    rsitelist_sorted_left, enamelist_sorted_left, _ = rsite_search_func(MCS, rsitelist, enamelist, starting_end=3)\n",
    "    for i, rsite in enumerate(rsitelist_sorted_left):\n",
    "        if not any(rsite in full_sequence for full_sequence in full_sequences) and not (rsite in backbone_no_MCS and rsite == rsite1):\n",
    "            rsite2 = rsite\n",
    "            ename2 = enamelist_sorted_left[i]\n",
    "            break\n",
    "        \n",
    "    return rsite1, ename1, rsite2, ename2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 469,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rsite_search(Gene, rsitelist, enamelist, modality, alpha, min_homology=0,\n",
    "                 left_of_Gene='', right_of_Gene='', FPGs=[], linker=''):\n",
    "\n",
    "    if modality not in [0, 3, 5]:\n",
    "        raise ValueError('Wrong starting_position value')\n",
    "\n",
    "    if modality == 5:\n",
    "\n",
    "        rsitelist_gene, enamelist_gene, rsite_position_list_gene = rsite_search_func(\n",
    "            Gene, rsitelist, enamelist, 5, min_homology)\n",
    "        rsitelist_5UTR, enamelist_5UTR, rsite_position_list_5TR = rsite_search_func(\n",
    "            left_of_Gene, rsitelist, enamelist, 3, min_homology)\n",
    "\n",
    "        start_end_sequences_gene = [generate_start_end_sequences(\n",
    "            left_of_Gene, Gene, 'right', rsite_pos, rsite, min_homology, alpha) for rsite_pos, rsite in zip(rsite_position_list_gene, rsitelist_gene)]\n",
    "        start_end_sequences_5UTR = [generate_start_end_sequences(\n",
    "            left_of_Gene, Gene, 'left', rsite_pos, rsite, min_homology, alpha) for rsite_pos, rsite in zip(rsite_position_list_5TR, rsitelist_5UTR)]\n",
    "\n",
    "        full_sequences = [\n",
    "            [start_seq+FPG+linker+end_seq for FPG in FPGs]\n",
    "            for start_seq, end_seq in start_end_sequences_gene+start_end_sequences_5UTR\n",
    "        ]\n",
    "\n",
    "        return_dict = {\n",
    "            'rsitelist': np.append(rsitelist_gene, rsitelist_5UTR),\n",
    "            'enamelist': np.append(enamelist_gene, enamelist_5UTR),\n",
    "            'rsite_position_list': np.append(rsite_position_list_gene, rsite_position_list_5TR),\n",
    "            'rsite_places': ['rsite in gene']*len(rsitelist_gene)+['rsite in 5`UTR']*len(rsitelist_5UTR),\n",
    "            'full_sequences': full_sequences,\n",
    "            'start_end_sequences': start_end_sequences_gene+start_end_sequences_5UTR \n",
    "        }\n",
    "\n",
    "    elif modality == 3:\n",
    "\n",
    "        rsitelist_gene, enamelist_gene, rsite_position_list_gene = rsite_search_func(\n",
    "            Gene, rsitelist, enamelist, 3, min_homology)\n",
    "        rsitelist_3UTR, enamelist_3UTR, rsite_position_list_3UTR = rsite_search_func(\n",
    "            right_of_Gene, rsitelist, enamelist, 5, min_homology)\n",
    "\n",
    "        start_end_sequences_gene = [generate_start_end_sequences(\n",
    "            Gene, right_of_Gene, 'left', rsite_pos, rsite, min_homology, alpha) for rsite_pos, rsite in zip(rsite_position_list_gene, rsitelist_gene)]\n",
    "        start_end_sequences_3UTR = [generate_start_end_sequences(\n",
    "            Gene, right_of_Gene, 'right', rsite_pos, rsite, min_homology, alpha) for rsite_pos, rsite in zip(rsite_position_list_3UTR, rsitelist_3UTR)]\n",
    "\n",
    "        full_sequences = [\n",
    "            [start_seq+linker+FPG+end_seq for FPG in FPGs]\n",
    "            for start_seq, end_seq in start_end_sequences_gene+start_end_sequences_3UTR\n",
    "        ]\n",
    "\n",
    "        return_dict = {\n",
    "            'rsitelist': np.append(rsitelist_gene, rsitelist_3UTR),\n",
    "            'enamelist': np.append(enamelist_gene, enamelist_3UTR),\n",
    "            'rsite_position_list': np.append(rsite_position_list_gene, rsite_position_list_3UTR),\n",
    "            'rsite_places': ['rsite in gene']*len(rsitelist_gene)+['rsite in 3`UTR']*len(rsitelist_3UTR),\n",
    "            'full_sequences': full_sequences,\n",
    "            'start_end_sequences': start_end_sequences_gene+start_end_sequences_3UTR\n",
    "        }\n",
    "\n",
    "    else:\n",
    "        rsitelist_5, enamelist_5, rsite_position_list_5 = rsite_search_func(left_of_Gene, rsitelist,\n",
    "                                                                            enamelist, 3, min_homology)\n",
    "        rsitelist_3, enamelist_3, rsite_position_list_3 = rsite_search_func(right_of_Gene, rsitelist,\n",
    "                                                                            enamelist, 5, min_homology)\n",
    "\n",
    "        start_end_sequences_5UTR = [generate_start_end_sequences(\n",
    "            left_of_Gene, Gene, 'left', rsite_pos, rsite, min_homology, alpha) for rsite_pos, rsite in zip(rsite_position_list_5, rsitelist_5)]\n",
    "        start_end_sequences_3UTR = [generate_start_end_sequences(\n",
    "            Gene, right_of_Gene, 'right', rsite_pos, rsite, min_homology, alpha) for rsite_pos, rsite in zip(rsite_position_list_3, rsitelist_3)]\n",
    "\n",
    "        full_sequences = [\n",
    "            [start_seq+end_seq]\n",
    "            for start_seq, end_seq in start_end_sequences_5UTR+start_end_sequences_3UTR\n",
    "        ]\n",
    "\n",
    "        return_dict = {\n",
    "            'rsitelist': np.append(rsitelist_5, rsitelist_3),\n",
    "            'enamelist': np.append(enamelist_5, enamelist_3),\n",
    "            'rsite_position_list': np.append(rsite_position_list_5, rsite_position_list_3),\n",
    "            'rsite_places': ['rsite in 5`']*len(rsitelist_5)+['rsite in 3`']*len(rsitelist_3),\n",
    "            'full_sequences': full_sequences,\n",
    "            'start_end_sequences': start_end_sequences_5UTR+start_end_sequences_3UTR\n",
    "        }\n",
    "\n",
    "    full_sequences_per_rsite = return_dict['full_sequences']\n",
    "    sort = np.argsort([len(x[0]) for x in full_sequences_per_rsite])\n",
    "    for key, value in return_dict.items():\n",
    "        return_dict[key] = list(np.array(value)[sort])\n",
    "\n",
    "    return return_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 470,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# INPUTS\n",
    "# 1. Plasmid backbone (pETUL): '*_backbone.fsa'\n",
    "# 2. Starting and ending index of MCS in the plasmid backbone (nucleotides counted from 1)\n",
    "# 3. min_homology: minimal homology in bp that should be included around the cutsite for integration into the yeast genome\n",
    "# 4. max_homology: maximal length in bp of the part used for the integration into the yeast genome \n",
    "# (this is used as a stopping criteria to search for cutsite, should be < 1000 bp).\n",
    "# 4. alpha: relative length of the homology part used for integration and homology part used for pop-out\n",
    "# 5. Gene to be tagged or deleted, including 1000 bp upstream of starting codon and 1000 bp downstream of stop: '*_gene.fsa'\n",
    "# 6. Linker sequence: '*_linker.fsa'\n",
    "# 7. A parameter that described the choice between deleting the gene (0), tagging at the 5' end (5) or tagging at the 3' end (3)\n",
    "# 8. In case of tagging, coding sequences of FPG (including the stop codon) should be given as a separate files ending with: '*_FPG' \n",
    "# (or maybe a single file with fasta, whatever is easier)\n",
    "# 9. A list of commercially available enzymes and their cutsites \n",
    "\n",
    "def main(backbone_path, MCS_start_ind, MCS_end_ind, min_homology, alpha, \n",
    "        Gene_path, linker_path, modality, enzyme_path, FPG_paths=[]):\n",
    "        \n",
    "    # Read all the files (1. and 2.)\n",
    "    _, backbone = read_from_fsa(backbone_path)\n",
    "    # print('Backbone len:', len(backbone))\n",
    "\n",
    "    MCS = backbone[MCS_start_ind-1:MCS_end_ind]\n",
    "    backbone_no_MCS_5 = backbone[0:MCS_start_ind-1]\n",
    "    backbone_no_MCS_3 = backbone[MCS_end_ind:]\n",
    "    # print('MCS len', len(MCS))\n",
    "\n",
    "    _, linker = read_from_fsa(linker_path)\n",
    "    # print('Linker len:', len(linker))\n",
    "    _, Gene_plus = read_from_fsa(Gene_path)\n",
    "    # print('Gene plus len:', len(Gene_plus))\n",
    "\n",
    "    left_of_Gene, Gene, right_of_Gene = read_gene_plus_string(Gene_plus)\n",
    "    # print('Gene len:', len(Gene))\n",
    "\n",
    "    FPGs = [read_from_fsa(path)[1] for path in FPG_paths] \n",
    "    # print('FPG len:', len(FPGs))\n",
    "\n",
    "    rsitelist, enamelist = read_enzyme_list(enzyme_path)\n",
    "    print(rsitelist[:10], enamelist[:10])\n",
    "\n",
    "\n",
    "    # 3.\n",
    "    rsite_dict = rsite_search(Gene, rsitelist, enamelist, modality, alpha,\n",
    "                              min_homology, left_of_Gene, right_of_Gene, FPGs, linker)\n",
    "\n",
    "    gene_rsitelist_sorted = rsite_dict['rsitelist']\n",
    "    gene_enamelist_sorted = rsite_dict['enamelist'] \n",
    "    gene_rsite_position_list_sorted = rsite_dict['rsite_position_list'] \n",
    "    rsite_places = rsite_dict['rsite_places'] \n",
    "    full_sequences_per_rsite = rsite_dict['full_sequences']\n",
    "    start_end_sequences = rsite_dict['start_end_sequences']\n",
    "    # print(gene_rsitelist_sorted, gene_enamelist_sorted, gene_rsite_position_list_sorted, rsite_places)\n",
    "\n",
    "    compatible_restriction_sites = []\n",
    "    optimal_plasmid = ''\n",
    "    MCS_rsites = []\n",
    "\n",
    "    # print([len(x[0]) for x in full_sequences_per_rsite])\n",
    "\n",
    "    for i in range(len(gene_rsitelist_sorted)):\n",
    "\n",
    "        print('\\n')\n",
    "\n",
    "        # 4. and 5.\n",
    "        rsite0 = gene_rsitelist_sorted[i]\n",
    "        rsite0_pos = gene_rsite_position_list_sorted[i]\n",
    "        ename0 = gene_enamelist_sorted[i]\n",
    "        rsite_place = rsite_places[i]\n",
    "        full_sequences = full_sequences_per_rsite[i]\n",
    "        start_seq, end_seq = start_end_sequences[i]\n",
    "        print(i, rsite0, ename0, 'side:',rsite_place)\n",
    "            \n",
    "        # check if all full sequences have only one rsite0  \n",
    "        if any([(full_sequence.count(rsite0) != 1) for full_sequence in full_sequences]):\n",
    "            print(\"\"\"Enzyme {} cannot satisfy the conditions.\n",
    "            rsite {} is not unique in the full sequence\"\"\".format(ename0, rsite0))\n",
    "            continue\n",
    "            \n",
    "\n",
    "        # 7.\n",
    "        # Search for the most outer CS from MCS that are not in 4. \n",
    "        rsite1, ename1, rsite2, ename2 = find_compatible_MCS_rsites(MCS, rsitelist, enamelist, full_sequences, backbone_no_MCS_5, backbone_no_MCS_3)\n",
    "\n",
    "        if not rsite1 or not rsite2:\n",
    "            print(\"\"\"Enzyme {} cannot satisfy the conditions.\n",
    "            good rsites not found in MCS\"\"\".format(ename0, rsite1, rsite2))\n",
    "            continue\n",
    "\n",
    "        # 8. \n",
    "        full_plasmid = assemble_plasmid(backbone_no_MCS_5, backbone_no_MCS_3, full_sequences[0])\n",
    "            \n",
    "        if full_plasmid.count(rsite0) != 1:\n",
    "            print(\"\"\"Enzyme {} cannot satisfy the conditions.\n",
    "            bad backbone for rsite0\"\"\".format(ename0, rsite0))\n",
    "\n",
    "        else:\n",
    "            if len(compatible_restriction_sites) == 0:\n",
    "                optimal_plasmid = full_plasmid\n",
    "                MCS_rsites = (rsite1, rsite2)\n",
    "            compatible_restriction_sites.append((ename0, rsite_place, ename1, ename2))\n",
    "            print(\"\"\"Enzyme {} can be used for cuttin the {}.\n",
    "            Enzymes that should be used for cloning inside the plasmid are {}, and {}.\"\"\".format(\n",
    "                ename0, rsite_place, ename1, ename2\n",
    "            ))\n",
    "            start_name = '5UTR' if modality in [5,0] else 'Gene end'\n",
    "            end_name = '3UTR' if modality in [3,0] else 'Gene start'\n",
    "            print(\"Pieces that should be used for cloning are {}:{}, {}:{}\".format(\n",
    "                start_name, start_seq, end_name, end_seq\n",
    "            ))\n",
    "\n",
    "    return optimal_plasmid, compatible_restriction_sites, MCS_rsites\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GACGTC' 'GGTACC' 'AACGTT' 'AGCGCT' 'CTTAAG' 'ACCGGT' 'GGGCCC' 'GTGCAC'\n",
      " 'GGCGCGCC' 'ATTAAT'] ['AatII, ZraI' 'Acc65I, KpnI, KpnI-HFÂ®' 'AclI' 'AfeI' 'AflII'\n",
      " 'AgeI, AgeI-HFÂ®' 'ApaI, PspOMI' 'ApaLI' 'AscI' 'AseI']\n",
      "\n",
      "\n",
      "0 TTTAAA DraI side: rsite in 3`UTR\n",
      "Enzyme DraI cannot satisfy the conditions.\n",
      "            rsite TTTAAA is not unique in the full sequence\n",
      "\n",
      "\n",
      "1 TTCGAA BstBI side: rsite in 3`UTR\n",
      "Enzyme BstBI cannot satisfy the conditions.\n",
      "            rsite TTCGAA is not unique in the full sequence\n",
      "\n",
      "\n",
      "2 ATTAAT AseI side: rsite in 3`UTR\n",
      "Enzyme AseI cannot satisfy the conditions.\n",
      "            bad backbone for rsite0\n",
      "\n",
      "\n",
      "3 TCCGGA BspEI side: rsite in 3`UTR\n",
      "Enzyme BspEI can be used for cuttin the rsite in 3`UTR.\n",
      "            Enzymes that should be used for cloning inside the plasmid are PaeR7I, XhoI, and EagI, EagI-HFÂ®.\n",
      "Pieces that should be used for cloning are Gene end:CTAAATATGAAGAGGTCTATTCTCCTAGTATAAAACATTTCGCATCAGAGACAGACGGTGCATGTACGGAAGATGAAATCAAAGAAGGGGAGAAATTCATTTTAAAGACATTGAAATTTAACCTAAATTATCCCAATCCGATGAATTTTCTGAGAAGAATTTCGAAAGCAGATGACTACGATATACAGTCTCGAACTCTTGCCAAATTCTTATTAGAGATATCATTGGTAGATTTCAGATTTATTGGGATACTACCCTCATTGTGTGCAGCAGCTGCGATGTTTATGTCGAGAAAAATGTTAGGTAAAGGTAAATGGGATGGAAATCTAATACACTATAGCGGCGGGTATACTAAAGAAGAACTTGCGCCCGTGTGTCACATGATAATGGATTATCTAGTGAGTCCAATTGTTCATGATGAATTTCATAGAAAATATCAATCTAGAAGATTTATGAAAGCTTCTATAATTTCCGTCCAATGGGCTTTAAAGGTTAGAAAAAACGGCTATGATATAATGACCTTGCATGAA, 3UTR:GAAGAGGGGCAGATGCTTAAAATATCTAAAACGATAATCGATAAATGTCCATATAAAATGTATGGCTGCTTTTTAATCATTTCTTTTCACATTTCCATTTATTCTTACAAATTTCACACCCAATTTTAGCGCCTTTCATTCTGTTATTTTTAAACTGACGATAAAAAATCATTCGAAGCCTAAGATTTCCAAAATAAAACTCCATTTTTATTCTCACTAATAACACCACACCATTAATTTTAATTTTTTAATTTTAATTTCAGATTTGTACCATAAGTTTTATTCTATTTCAGTTCAACCGTTTCCGGATGCATGTAGCGAGGGTAGACAAAGAAAAGCATCAGTATCGGATGTTGATATTATGGGAAATATCATAACC\n",
      "\n",
      "\n",
      "4 ATGCAT NsiI, NsiI-HFÂ® side: rsite in 3`UTR\n",
      "Enzyme NsiI, NsiI-HFÂ® cannot satisfy the conditions.\n",
      "            bad backbone for rsite0\n",
      "\n",
      "\n",
      "5 TTATAA PsiI side: rsite in 3`UTR\n",
      "Enzyme PsiI cannot satisfy the conditions.\n",
      "            bad backbone for rsite0\n",
      "\n",
      "\n",
      "6 AAGCTT HindIII, HindIII-HFÂ® side: rsite in gene\n",
      "Enzyme HindIII, HindIII-HFÂ® can be used for cuttin the rsite in gene.\n",
      "            Enzymes that should be used for cloning inside the plasmid are PaeR7I, XhoI, and EagI, EagI-HFÂ®.\n",
      "Pieces that should be used for cloning are Gene end:CCATTAACATAATGGACAGGTTTTTAGGTAAAGAGCTAGTTCAACTGGATAAGTTACAATTGGTTGGCACATCATGCCTTTTCATTGCCTCTAAATATGAAGAGGTCTATTCTCCTAGTATAAAACATTTCGCATCAGAGACAGACGGTGCATGTACGGAAGATGAAATCAAAGAAGGGGAGAAATTCATTTTAAAGACATTGAAATTTAACCTAAATTATCCCAATCCGATGAATTTTCTGAGAAGAATTTCGAAAGCAGATGACTACGATATACAGTCTCGAACTCTTGCCAAATTCTTATTAGAGATATCATTGGTAGATTTCAGATTTATTGGGATACTACCCTCATTGTGTGCAGCAGCTGCGATGTTTATGTCGAGAAAAATGTTAGGTAAAGGTAAATGGGATGGAAATCTAATACACTATAGCGGCGGGTATACTAAAGAAGAACTTGCGCCCGTGTGTCACATGATAATGGATTATCTAGTGAGTCCAATTGTTCATGATGAATTTCATAGAAAATATCAATCTAGAAGATTTATGAAAGCTTCTATAATTTCCGTCCAATGGGCTTTAAAGGTTAGAAAAAACGGCTATGATATAATGACCTTGCATGAA, 3UTR:GAAGAGGGGCAGATGCTTAAAATATCTAAAACGATAATCGATAAATGTCCATATAAAATGTATGGCTGCTTTTTAATCATTTCTTTTCACATTTCCATTTATTCTTACAAATTTCACACCCAATTTTAGCGCCTTTCATTCTGTTATTTTTAAACTGACGATAAAAAATCATTCGAAGCCTAAGATTTCCAAAATAAAACTCCATTTTTATTCTCACTAATAACACCACACCATTAATTTTAATTTTTTAATTTTAATTTCAGATTTGTACCATAAGTTTTATTCTATTTCAGTTCAACCGTTTCCGGATGCATGTAGCGAGGGTAGACAAAGAAAAGCATCAGTATCGGATGTTGATATTATGGGAAATATCATAACCTTAGTCTCTTAAAAAATGCTGCTTTCTTCTCATTTTTCATTTCTTTTAAAATTATTTCAGTATAGTTATTATAATTAACATTGGAATAACATTAATAATAATAATTTTTTAATAGTAATAATGATAATAATAATAATAAAAATAAAAGAATAAATTTAAAAACAAAAAAAATTCTAATATTAATGATGATAATAGTAGTAATACTGGTGGTATTATATATGTTGAAGTGAACACGCATATCACTATAAACTATATATACATATATATAGATAGATAGATAGATGCAAAAAATTTTTTCCTTTTAGTTCAGCAAAAAGAAAAGAAAATGTAAAGAGTATGCGAATTCATGAGCATTACTAGTACTAATCTACTTAAGATTAAATAGATTTTGAAAGTTGCTATGCATTTCGGATGTACACCACTTGAAAGTTTGGAAATAAACAGAGCCATACCTTGGTTTCTTGTACTTCAAAATCAATGAATCTAAAT\n",
      "\n",
      "\n",
      "7 AATATT SspI, SspI-HFÂ® side: rsite in 3`UTR\n",
      "Enzyme SspI, SspI-HFÂ® cannot satisfy the conditions.\n",
      "            rsite AATATT is not unique in the full sequence\n",
      "\n",
      "\n",
      "8 TCTAGA XbaI side: rsite in gene\n",
      "Enzyme XbaI can be used for cuttin the rsite in gene.\n",
      "            Enzymes that should be used for cloning inside the plasmid are PaeR7I, XhoI, and EagI, EagI-HFÂ®.\n",
      "Pieces that should be used for cloning are Gene end:GAGACTTTGTATCTTGCCATTAACATAATGGACAGGTTTTTAGGTAAAGAGCTAGTTCAACTGGATAAGTTACAATTGGTTGGCACATCATGCCTTTTCATTGCCTCTAAATATGAAGAGGTCTATTCTCCTAGTATAAAACATTTCGCATCAGAGACAGACGGTGCATGTACGGAAGATGAAATCAAAGAAGGGGAGAAATTCATTTTAAAGACATTGAAATTTAACCTAAATTATCCCAATCCGATGAATTTTCTGAGAAGAATTTCGAAAGCAGATGACTACGATATACAGTCTCGAACTCTTGCCAAATTCTTATTAGAGATATCATTGGTAGATTTCAGATTTATTGGGATACTACCCTCATTGTGTGCAGCAGCTGCGATGTTTATGTCGAGAAAAATGTTAGGTAAAGGTAAATGGGATGGAAATCTAATACACTATAGCGGCGGGTATACTAAAGAAGAACTTGCGCCCGTGTGTCACATGATAATGGATTATCTAGTGAGTCCAATTGTTCATGATGAATTTCATAGAAAATATCAATCTAGAAGATTTATGAAAGCTTCTATAATTTCCGTCCAATGGGCTTTAAAGGTTAGAAAAAACGGCTATGATATAATGACCTTGCATGAA, 3UTR:GAAGAGGGGCAGATGCTTAAAATATCTAAAACGATAATCGATAAATGTCCATATAAAATGTATGGCTGCTTTTTAATCATTTCTTTTCACATTTCCATTTATTCTTACAAATTTCACACCCAATTTTAGCGCCTTTCATTCTGTTATTTTTAAACTGACGATAAAAAATCATTCGAAGCCTAAGATTTCCAAAATAAAACTCCATTTTTATTCTCACTAATAACACCACACCATTAATTTTAATTTTTTAATTTTAATTTCAGATTTGTACCATAAGTTTTATTCTATTTCAGTTCAACCGTTTCCGGATGCATGTAGCGAGGGTAGACAAAGAAAAGCATCAGTATCGGATGTTGATATTATGGGAAATATCATAACCTTAGTCTCTTAAAAAATGCTGCTTTCTTCTCATTTTTCATTTCTTTTAAAATTATTTCAGTATAGTTATTATAATTAACATTGGAATAACATTAATAATAATAATTTTTTAATAGTAATAATGATAATAATAATAATAAAAATAAAAGAATAAATTTAAAAACAAAAAAAATTCTAATATTAATGATGATAATAGTAGTAATACTGGTGGTATTATATATGTTGAAGTGAACACGCATATCACTATAAACTATATATACATATATATAGATAGATAGATAGATGCAAAAAATTTTTTCCTTTTAGTTCAGCAAAAAGAAAAGAAAATGTAAAGAGTATGCGAATTCATGAGCATTACTAGTACTAATCTACTTAAGATTAAATAGATTTTGAAAGTTGCTATGCATTTCGGATGTACACCACTTGAAAGTTTGGAAATAAACAGAGCCATACCTTGGTTTCTTGTACTTCAAAATCAATGAATCTAAATGAGTTTTGGAACTAGCGATGTC\n",
      "\n",
      "\n",
      "9 TCATGA BspHI side: rsite in gene\n",
      "Enzyme BspHI cannot satisfy the conditions.\n",
      "            rsite TCATGA is not unique in the full sequence\n",
      "\n",
      "\n",
      "10 CAATTG MfeI, MfeI-HFÂ® side: rsite in gene\n",
      "Enzyme MfeI, MfeI-HFÂ® cannot satisfy the conditions.\n",
      "            rsite CAATTG is not unique in the full sequence\n",
      "\n",
      "\n",
      "11 GTATAC BstZ17I, BstZ17I-HFÂ® side: rsite in gene\n",
      "Enzyme BstZ17I, BstZ17I-HFÂ® cannot satisfy the conditions.\n",
      "            bad backbone for rsite0\n",
      "\n",
      "\n",
      "12 CAGCTG PvuII, PvuII-HFÂ® side: rsite in gene\n",
      "Enzyme PvuII, PvuII-HFÂ® cannot satisfy the conditions.\n",
      "            rsite CAGCTG is not unique in the full sequence\n",
      "\n",
      "\n",
      "13 GATATC EcoRV, EcoRV-HFÂ® side: rsite in gene\n",
      "Enzyme EcoRV, EcoRV-HFÂ® cannot satisfy the conditions.\n",
      "            rsite GATATC is not unique in the full sequence\n",
      "\n",
      "\n",
      "14 GAATTC EcoRI, EcoRI-HFÂ® side: rsite in 3`UTR\n",
      "Enzyme EcoRI, EcoRI-HFÂ® cannot satisfy the conditions.\n",
      "            rsite GAATTC is not unique in the full sequence\n",
      "\n",
      "\n",
      "15 TTCGAA BstBI side: rsite in gene\n",
      "Enzyme BstBI cannot satisfy the conditions.\n",
      "            rsite TTCGAA is not unique in the full sequence\n",
      "\n",
      "\n",
      "16 TCATGA BspHI side: rsite in 3`UTR\n",
      "Enzyme BspHI cannot satisfy the conditions.\n",
      "            rsite TCATGA is not unique in the full sequence\n",
      "\n",
      "\n",
      "17 ACTAGT SpeI, SpeI-HFÂ® side: rsite in 3`UTR\n",
      "Enzyme SpeI, SpeI-HFÂ® can be used for cuttin the rsite in 3`UTR.\n",
      "            Enzymes that should be used for cloning inside the plasmid are PaeR7I, XhoI, and EagI, EagI-HFÂ®.\n",
      "Pieces that should be used for cloning are Gene end:ACTTCTATAAAGGAGGGGAGAGCTGAGCTGCCTGCAAATATAAGTTTACAAGAATCCTCCTCAGCGAAGGAGATAATCCAGCATGATCCCCTAAAAGGCGTTGGATCAAGCACTGAGGTAGTCCATAACTCGGTAGAAAACGAAAAACTTCATCCAGCTAGAAGTCAACTTCAAGTTAGAAATACCGAAAGTGAAACTGATAGTGGAAAAAAAAGACCAATTTCTACAATTGTTGAACAAGAACTGCCCAAAAAGTTTAAAGTGTGCGATGAAAATGGCAAAGAAGAATATGAATGGGAAGACCTAGATGCAGAAGATGTAAATGATCCATTCATGGTCAGCGAGTACGTCAATGATATATTCGAATATCTCCACCAACTAGAGGTCATTACTCTTCCAAAGAAGGAAGATCTCTATCAGCATAGAAATATTCATCAAAATCGAGATATCCTAGTTAATTGGTTGGTTAAAATCCATAATAAATTCGGCTTATTACCGGAGACTTTGTATCTTGCCATTAACATAATGGACAGGTTTTTAGGTAAAGAGCTAGTTCAACTGGATAAGTTACAATTGGTTGGCACATCATGCCTTTTCATTGCCTCTAAATATGAAGAGGTCTATTCTCCTAGTATAAAACATTTCGCATCAGAGACAGACGGTGCATGTACGGAAGATGAAATCAAAGAAGGGGAGAAATTCATTTTAAAGACATTGAAATTTAACCTAAATTATCCCAATCCGATGAATTTTCTGAGAAGAATTTCGAAAGCAGATGACTACGATATACAGTCTCGAACTCTTGCCAAATTCTTATTAGAGATATCATTGGTAGATTTCAGATTTATTGGGATACTACCCTCATTGTGTGCAGCAGCTGCGATGTTTATGTCGAGAAAAATGTTAGGTAAAGGTAAATGGGATGGAAATCTAATACACTATAGCGGCGGGTATACTAAAGAAGAACTTGCGCCCGTGTGTCACATGATAATGGATTATCTAGTGAGTCCAATTGTTCATGATGAATTTCATAGAAAATATCAATCTAGAAGATTTATGAAAGCTTCTATAATTTCCGTCCAATGGGCTTTAAAGGTTAGAAAAAACGGCTATGATATAATGACCTTGCATGAA, 3UTR:GAAGAGGGGCAGATGCTTAAAATATCTAAAACGATAATCGATAAATGTCCATATAAAATGTATGGCTGCTTTTTAATCATTTCTTTTCACATTTCCATTTATTCTTACAAATTTCACACCCAATTTTAGCGCCTTTCATTCTGTTATTTTTAAACTGACGATAAAAAATCATTCGAAGCCTAAGATTTCCAAAATAAAACTCCATTTTTATTCTCACTAATAACACCACACCATTAATTTTAATTTTTTAATTTTAATTTCAGATTTGTACCATAAGTTTTATTCTATTTCAGTTCAACCGTTTCCGGATGCATGTAGCGAGGGTAGACAAAGAAAAGCATCAGTATCGGATGTTGATATTATGGGAAATATCATAACCTTAGTCTCTTAAAAAATGCTGCTTTCTTCTCATTTTTCATTTCTTTTAAAATTATTTCAGTATAGTTATTATAATTAACATTGGAATAACATTAATAATAATAATTTTTTAATAGTAATAATGATAATAATAATAATAAAAATAAAAGAATAAATTTAAAAACAAAAAAAATTCTAATATTAATGATGATAATAGTAGTAATACTGGTGGTATTATATATGTTGAAGTGAACACGCATATCACTATAAACTATATATACATATATATAGATAGATAGATAGATGCAAAAAATTTTTTCCTTTTAGTTCAGCAAAAAGAAAAGAAAATGTAAAGAGTATGCGAATTCATGAGCATTACTAGTACTAATCTACTTAAGATTAAATAGATTTTGAAAGTTGCTATGCATTTCGGATGTACACCACTTGAAAGTT\n",
      "\n",
      "\n",
      "18 AGTACT ScaI-HFÂ® side: rsite in 3`UTR\n",
      "Enzyme ScaI-HFÂ® cannot satisfy the conditions.\n",
      "            bad backbone for rsite0\n",
      "\n",
      "\n",
      "19 TTTAAA DraI side: rsite in gene\n",
      "Enzyme DraI cannot satisfy the conditions.\n",
      "            rsite TTTAAA is not unique in the full sequence\n",
      "\n",
      "\n",
      "20 CTTAAG AflII side: rsite in 3`UTR\n",
      "Enzyme AflII cannot satisfy the conditions.\n",
      "            bad backbone for rsite0\n",
      "\n",
      "\n",
      "21 TGTACA BsrGI, BsrGI-HFÂ® side: rsite in 3`UTR\n",
      "Enzyme BsrGI, BsrGI-HFÂ® cannot satisfy the conditions.\n",
      "            rsite TGTACA is not unique in the full sequence\n",
      "\n",
      "\n",
      "22 AATATT SspI, SspI-HFÂ® side: rsite in gene\n",
      "Enzyme SspI, SspI-HFÂ® cannot satisfy the conditions.\n",
      "            rsite AATATT is not unique in the full sequence\n",
      "\n",
      "\n",
      "23 AGATCT BglII side: rsite in gene\n",
      "Enzyme BglII can be used for cuttin the rsite in gene.\n",
      "            Enzymes that should be used for cloning inside the plasmid are PaeR7I, XhoI, and EagI, EagI-HFÂ®.\n",
      "Pieces that should be used for cloning are Gene end:TAGAGAAAGCGTATCAAGATCAACAGCCGCACAAGAGGAGAAAAGAACCCTGAAAGAAAATGGTATCCAACTCCCCAAAAACAATCTTTTAGATGATAAAGAAAACCAAGACCCAAGTAGTCAGCAATTTGGTGCGCTAACTTCTATAAAGGAGGGGAGAGCTGAGCTGCCTGCAAATATAAGTTTACAAGAATCCTCCTCAGCGAAGGAGATAATCCAGCATGATCCCCTAAAAGGCGTTGGATCAAGCACTGAGGTAGTCCATAACTCGGTAGAAAACGAAAAACTTCATCCAGCTAGAAGTCAACTTCAAGTTAGAAATACCGAAAGTGAAACTGATAGTGGAAAAAAAAGACCAATTTCTACAATTGTTGAACAAGAACTGCCCAAAAAGTTTAAAGTGTGCGATGAAAATGGCAAAGAAGAATATGAATGGGAAGACCTAGATGCAGAAGATGTAAATGATCCATTCATGGTCAGCGAGTACGTCAATGATATATTCGAATATCTCCACCAACTAGAGGTCATTACTCTTCCAAAGAAGGAAGATCTCTATCAGCATAGAAATATTCATCAAAATCGAGATATCCTAGTTAATTGGTTGGTTAAAATCCATAATAAATTCGGCTTATTACCGGAGACTTTGTATCTTGCCATTAACATAATGGACAGGTTTTTAGGTAAAGAGCTAGTTCAACTGGATAAGTTACAATTGGTTGGCACATCATGCCTTTTCATTGCCTCTAAATATGAAGAGGTCTATTCTCCTAGTATAAAACATTTCGCATCAGAGACAGACGGTGCATGTACGGAAGATGAAATCAAAGAAGGGGAGAAATTCATTTTAAAGACATTGAAATTTAACCTAAATTATCCCAATCCGATGAATTTTCTGAGAAGAATTTCGAAAGCAGATGACTACGATATACAGTCTCGAACTCTTGCCAAATTCTTATTAGAGATATCATTGGTAGATTTCAGATTTATTGGGATACTACCCTCATTGTGTGCAGCAGCTGCGATGTTTATGTCGAGAAAAATGTTAGGTAAAGGTAAATGGGATGGAAATCTAATACACTATAGCGGCGGGTATACTAAAGAAGAACTTGCGCCCGTGTGTCACATGATAATGGATTATCTAGTGAGTCCAATTGTTCATGATGAATTTCATAGAAAATATCAATCTAGAAGATTTATGAAAGCTTCTATAATTTCCGTCCAATGGGCTTTAAAGGTTAGAAAAAACGGCTATGATATAATGACCTTGCATGAA, 3UTR:GAAGAGGGGCAGATGCTTAAAATATCTAAAACGATAATCGATAAATGTCCATATAAAATGTATGGCTGCTTTTTAATCATTTCTTTTCACATTTCCATTTATTCTTACAAATTTCACACCCAATTTTAGCGCCTTTCATTCTGTTATTTTTAAACTGACGATAAAAAATCATTCGAAGCCTAAGATTTCCAAAATAAAACTCCATTTTTATTCTCACTAATAACACCACACCATTAATTTTAATTTTTTAATTTTAATTTCAGATTTGTACCATAAGTTTTATTCTATTTCAGTTCAACCGTTTCCGGATGCATGTAGCGAGGGTAGACAAAGAAAAGCATCAGTATCGGATGTTGATATTATGGGAAATATCATAACCTTAGTCTCTTAAAAAATGCTGCTTTCTTCTCATTTTTCATTTCTTTTAAAATTATTTCAGTATAGTTATTATAATTAACATTGGAATAACATTAATAATAATAATTTTTTAATAGTAATAATGATAATAATAATAATAAAAATAAAAGAATAAATTTAAAAACAAAAAAAATTCTAATATTAATGATGATAATAGTAGTAATACTGGTGGTATTATATATGTTGAAGTGAACACGCATATCACTATAAACTATATATACATATATATAGATAGATAGATAGATGCAAAAAATTTTTTCCTTTTAGTTCAGCAAAAAGAAAAGAAAATGTAAAGAGTATGCGAATTCATGAGCATTACTAGTACTAATCTACTTAAGATTAAATAGATTTTGAAAGTTGCTATGCATTTCGGATGTACACCACTTGAAAGTTTGGAAATAAACAGAGCCATACCTTGGTTTCTTGTACTTCAAAATCAATGAATCTAAATGAGTTTTGGAACTAGCGATGTCTTTGACTAGATCAATGCAGAGAGACTGAAACGCTTCATCGTGTATTGGATCGATACCACCACTGTAATGCTGTAGTGTTCCATTCCATAGCTCGTTCTTGTTTCTGTTGG\n",
      "\n",
      "\n",
      "24 ATCGAT BspDI, ClaI side: rsite in 3`UTR\n",
      "Enzyme BspDI, ClaI cannot satisfy the conditions.\n",
      "            rsite ATCGAT is not unique in the full sequence\n",
      "8167 ('CTCGAG', 'CGGCCG')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('BspEI', 'rsite in 3`UTR', 'PaeR7I, XhoI', 'EagI, EagI-HFÂ®'),\n",
       " ('HindIII, HindIII-HFÂ®', 'rsite in gene', 'PaeR7I, XhoI', 'EagI, EagI-HFÂ®'),\n",
       " ('XbaI', 'rsite in gene', 'PaeR7I, XhoI', 'EagI, EagI-HFÂ®'),\n",
       " ('SpeI, SpeI-HFÂ®', 'rsite in 3`UTR', 'PaeR7I, XhoI', 'EagI, EagI-HFÂ®'),\n",
       " ('BglII', 'rsite in gene', 'PaeR7I, XhoI', 'EagI, EagI-HFÂ®')]"
      ]
     },
     "execution_count": 471,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "backbone_path = './Test_examples/pETUL_backbone.fsa'\n",
    "linker_path = './Test_examples/long_linker.fsa'\n",
    "\n",
    "Gene_path = './Test_examples/CLB2_3p_labeling/CLB2_pm_1000.fsa'\n",
    "FPG_paths = [\n",
    "    './Test_examples/CLB2_3p_labeling/mCherry_FPG.fsa',\n",
    "    './Test_examples/CLB2_3p_labeling/ymNeonGreen_FPG.fsa',\n",
    "    './Test_examples/CLB2_3p_labeling/ymTq2_FPG.fsa'\n",
    "]\n",
    "modality = 3\n",
    "\n",
    "# Gene_path = './Test_examples/CLB5_3p_labeling/CLB5_pm_1000_gene.fsa'\n",
    "# FPG_paths = [\n",
    "#     './Test_examples/CLB5_3p_labeling/mCherry_FPG.fsa',\n",
    "#     './Test_examples/CLB5_3p_labeling/ymNeonGreen_FPG.fsa',\n",
    "#     './Test_examples/CLB5_3p_labeling/ymTq2_FPG.fsa'\n",
    "# ]\n",
    "# gene_side = 5\n",
    "\n",
    "# Gene_path = './Test_examples/GAL4_deletion/GAL4_pm_1000_gene.fsa'\n",
    "# FPG_paths = ['']\n",
    "# modality = 0\n",
    "\n",
    "# Gene_path = './Test_examples/HTB2_3p_labeling/HTB2_pm_1000.fsa'\n",
    "# FPG_paths = [\n",
    "#     './Test_examples/HTB2_3p_labeling/mCherry_FPG.fsa',\n",
    "#     './Test_examples/HTB2_3p_labeling/sfGFP_FPG.fsa'\n",
    "# ]\n",
    "# gene_side = 5\n",
    "\n",
    "enzymefile = './Chris_code/raw_enzyme_list.txt'\n",
    "\n",
    "optimal_plasmid, compatible_restriction_sites, mcs = main(backbone_path, 1, 108, 70, 1.4, Gene_path, linker_path, \n",
    "                                    modality, enzymefile, FPG_paths)\n",
    "\n",
    "print(len(optimal_plasmid), mcs)\n",
    "compatible_restriction_sites\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 465,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_, backbone = read_from_fsa('./Test_examples/pETUL_backbone.fsa')\n",
    "MCS = backbone[0:108]\n",
    "MCS.find('GGTACC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "105"
      ]
     },
     "execution_count": 338,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "read_from_fsa('./Test_examples/CLB2_3p_labeling/ymNeonGreen_FPG.fsa')[1].find('GGTACC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def dna_to_protein(dna):\n",
    "     #Translates DNA into a Protein. it truncates the 3' tail that doesn't make a full codon\n",
    "     #STOP is denoted by '*'\n",
    "     \n",
    "\tdna = dna.upper()\n",
    "\n",
    "\tgenetic_code = {\n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                 \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',\n",
    "    }\n",
    "\tprotein = ''\n",
    "\n",
    "\n",
    "\tif(len(dna)%3 != 0):\n",
    "\t\tdna = dna[: -(len(dna)%3)]\n",
    "\n",
    "\tfor i in range(0, len(dna), 3):\n",
    "\t\tcode = dna[i:i+3]\n",
    "\t\tif(code in genetic_code.keys()):\n",
    "\t\t\tprotein += genetic_code[code]\n",
    "\t\telse:\n",
    "\t\t\tprotein = ''\n",
    "\t\t\tbreak\n",
    "\t\n",
    "\treturn protein\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def  find_additional_cutsites(plasmid, rsitelist, enamelist):\n",
    "#this function finds all appropropriate cutsites that could be added between different pieces of the insert in the plasmid (e.g. between the gene and the linker or between the linker and the FPG etc)\n",
    "#a good cutsite has to fullfill three criteria:\n",
    "#1. Divisible by 3 (this could be circuimvented by accomidating the linker length)\n",
    "#2. Does not cut the final plasmid\n",
    "#3. Does not introduce a stop codon\n",
    "\n",
    "    good_rsite_list = []\n",
    "    good_enzyme_list = []\n",
    "    for rsite, enzyme in zip(rsitelist, enamelist):\n",
    "        if len(rsite)%3 == 0 and plasmid.count(rsite) == 0 and dna_to_protein(rsite).count('*') == 0:\n",
    "            good_rsite_list.append(rsite)\n",
    "            good_enzyme_list.append(enzyme)\n",
    "\n",
    "    return good_rsite_list, good_enzyme_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_to_protein('TGCTACtga').count('*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "841159ca92e470d2c15efac886a071402c9a594398da0f615defd04595a3ae06"
  },
  "kernelspec": {
   "display_name": "Python 3.6.5 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
